Add support for "classifier" dependency substitution (#13142)
gradle/gradle: subprojects/docs/src/docs/userguide/dep-man/03-controlling-transitive-dependencies/resolution_rules.adoc

Note that a project that is substituted must be included in the multi-project build (via `settings.gradle`).
Dependency substitution rules take care of replacing the module dependency with the project dependency, but do not implicitly include the project in the build.

[[sec:substitution_with_classifier]]
=== Substituting a dependency with a classifier or artifact

While external modules are in general addressed via their group/artifact/version coordinates, it is common that such modules are published with additional artifacts that you may want to use in place of the main artifact.

> This is typically the case for _classified_ artifacts, but you may also need to select an artifact with a different file type or extension.

Reviewer: I wonder myself what the practical differente between _file type_ and _extension_ is?

Author: I'm not quite sure either, it seems to be mostly relevant for Ivy.

Reviewer 2: In Ivy, it allows you to do something like `type=docs`, `ext=jar` for the sources and javadoc jars for example. These artifacts could then be located in a different folder on the server. Of course not really with Gradle since we do not fully support `type` 😉

> Gradle discourages use of classifiers in dependencies and prefers to model such artifacts as <<variant_model.adoc#understanding-variant-selection,variants of the same thing>>.

Reviewer: ```suggestion
Gradle discourages use of classifiers in dependencies and prefers to model such artifacts as <<variant_model.adoc#understanding-variant-selection,variants additional variants of a module>>.
```

There are lots of advantages of using variants instead of classified artifacts, including, but not only, a different set of dependencies for those artifacts.

> However, in order to help bridging the two models, Gradle provides means to select a different classifier in a substitution rule, or, on the contrary, replace a dependency which uses a classifier with a dependency which doesn't.

Reviewer: ```suggestion
However, in order to help bridging the two models, Gradle provides means to change or remove a classifier in a substitution rule.
```
